In [1]:
#Import libraries
import pandas as pd

In [3]:
#Read the parquet file
compliments_df = pd.read_parquet('../data/raw/compliments.parquet')

In [5]:
#Find the first 10 rows of the dataset
compliments_df.head(10)

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
0,[xx:compliments],Compliments,0055742346022,"[en:canada, en:france]","[{'key': '4', 'imgid': None, 'rev': None, 'siz...","[en:crust, en:filling, en:fortified-wheat-flou...","[en:crust, en:filling, en:fortified-wheat-flou...","[{""percent_max"":100.0,""percent_min"":50.0,""is_i...","[{'name': 'added-sugars', 'value': None, '100g...","[{'lang': 'main', 'text': 'Strawberry Cereal B...",g,300.0,300 g,[en:eggs],[]
1,[xx:compliments],Compliments,0055742348279,[en:canada],"[{'key': '9', 'imgid': None, 'rev': None, 'siz...","[en:peanut, en:sugar, en:hydrogenated-vegetabl...","[en:peanut, en:nut, en:sugar, en:added-sugar, ...","[{""percent_max"":100.0,""percent_min"":93.9,""is_i...","[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Smooth Peanut Butte...",g,500.0,500 g,[en:nuts],[]
2,[xx:compliments],Compliments,0055742350500,[en:france],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...","[en:peanut, fr:sucre-a-glacer, en:hydrogenated...","[en:peanut, en:nut, fr:sucre-a-glacer, en:hydr...","[{""percent_max"":100.0,""percent_min"":20.0,""is_i...","[{'name': 'energy', 'value': None, '100g': 270...","[{'lang': 'main', 'text': 'Beurre d'arachide (...",None,None,,[],[]
3,[xx:compliments],Compliments,0055742357912,"[en:canada, en:france]","[{'key': '3', 'imgid': None, 'rev': None, 'siz...","[en:durum-wheat-semolina, en:e375, en:ferrous-...","[en:durum-wheat-semolina, en:cereal, en:wheat,...","[{""percent_max"":100.0,""percent_min"":16.6666666...","[{'name': 'added-sugars', 'value': None, '100g...","[{'lang': 'main', 'text': 'Fettucine'}, {'lang...",g,900.0,900g,[],"[en:niacin, en:riboflavin, en:thiamin-mononitr..."
4,[xx:compliments],Compliments,0055742358988,[en:france],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Melasse'}, {'lang':...",None,None,None,[],[]
5,[xx:compliments],Compliments,0055742359886,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...","[fr:not-a-significant-source-of-saturated-fat,...","[fr:not-a-significant-source-of-saturated-fat,...","[{""percent_max"":100.0,""percent_min"":6.25,""is_i...","[{'name': 'added-sugars', 'value': None, '100g...","[{'lang': 'main', 'text': 'Persil Déshydraté E...",None,None,None,[],[en:vitamin-a]
6,[xx:compliments],Compliments,0055742363425,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 0.620...","[{'lang': 'main', 'text': 'Crushed tomatoes'},...",ml,796.0,796 mL,[],[]
7,[xx:compliments],Compliments,0055742366891,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",[en:honey],"[en:honey, en:added-sugar]","[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'galactose', 'value': None, '100g': ...","[{'lang': 'main', 'text': 'Miel biologique pur...",g,500,500 g,[],[]
8,[xx:compliments],Compliments,0055742367164,[en:france],"[{'key': '4', 'imgid': None, 'rev': None, 'siz...",[fr:peanuts],[fr:peanuts],"[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'fat', 'value': None, '100g': 46.666...","[{'lang': 'main', 'text': 'Natural crunchy pea...",None,None,None,[],[]
9,[xx:compliments],Compliments,0055742369212,[en:canada],"[{'key': '4', 'imgid': None, 'rev': None, 'siz...",[en:date],"[en:date, en:fruit]","[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'beta-carotene', 'value': None, '100...","[{'lang': 'main', 'text': 'Pitted dates'}, {'l...",g,600,600g,[],[]


In [7]:
#Find the number of null values for each column
compliments_df.isnull().sum()

brands_tags                     0
brands                          0
code                            0
countries_tags                  6
images                          0
ingredients_original_tags    1363
ingredients_tags             1363
ingredients                  1363
nutriments                     77
product_name                    0
product_quantity_unit        1389
product_quantity             1376
quantity                     1324
traces_tags                     0
vitamins_tags                1220
dtype: int64

In [9]:
# Find percentage of null values for each column
compliments_df.isnull().mean() * 100

brands_tags                   0.000000
brands                        0.000000
code                          0.000000
countries_tags                0.326975
images                        0.000000
ingredients_original_tags    74.277929
ingredients_tags             74.277929
ingredients                  74.277929
nutriments                    4.196185
product_name                  0.000000
product_quantity_unit        75.694823
product_quantity             74.986376
quantity                     72.152589
traces_tags                   0.000000
vitamins_tags                66.485014
dtype: float64

In [11]:
#Get the columns in the dataset
compliments_df.columns

Index(['brands_tags', 'brands', 'code', 'countries_tags', 'images',
       'ingredients_original_tags', 'ingredients_tags', 'ingredients',
       'nutriments', 'product_name', 'product_quantity_unit',
       'product_quantity', 'quantity', 'traces_tags', 'vitamins_tags'],
      dtype='object')

In [13]:
#Find the number of rows and columns in the dataset
compliments_df.shape

(1835, 15)

In [15]:
#Find the datatype of each column in the dataset
for i in compliments_df.columns:
    print(compliments_df[i].dtype)

object
object
object
object
object
object
object
object
object
object
object
object
object
object
object


In [17]:
#Since brands_tags itself is like an array, explode() has been used to flatten and obtain the unique values
compliments_df['brands_tags'].explode().unique()

array(['xx:compliments', 'compliments', 'xx:Compliments'], dtype=object)

In [21]:
#Obtain rows which matches brands_tags exactly with 'compliments'
compliments_only = compliments_df[compliments_df['brands_tags'].apply(lambda x: len(x) == 1 and x[0] == 'compliments')]
len(compliments_only)

113

In [23]:
#Obtain those products where the brand name was mentioned as 'compliments'
compliments_only[['brands_tags', 'product_name']]

,brands_tags,product_name
11,[compliments],"[{'lang': 'main', 'text': 'Brown Rice'}, {'lan..."
15,[compliments],"[{'lang': 'main', 'text': 'Craquelins de blé m..."
18,[compliments],"[{'lang': 'main', 'text': 'Gluten-Free Crisp R..."
44,[compliments],"[{'lang': 'main', 'text': 'Quick oats'}, {'lan..."
52,[compliments],"[{'lang': 'main', 'text': 'Medium Cheddar'}, {..."
...,...,...
1411,[compliments],"[{'lang': 'main', 'text': 'Mild pepperoni'}, {..."
1419,[compliments],"[{'lang': 'main', 'text': 'English Muffins'}, ..."
1425,[compliments],"[{'lang': 'main', 'text': 'Bacon entièrement c..."
1438,[compliments],"[{'lang': 'main', 'text': 'Crevettes'}, {'lang..."


In [25]:
#Obtain the brands_tags and product_name from the dataset where the countries_tags is null
compliments_df.loc[compliments_df['countries_tags'].isnull(), ['brands_tags', 'product_name']]

,brands_tags,product_name
240,[compliments],"[{'lang': 'main', 'text': 'Prepared horseradis..."
429,[compliments],"[{'lang': 'main', 'text': 'Whole Wheat Naan Ro..."
1273,[xx:compliments],"[{'lang': 'main', 'text': 'Granulated Sweetene..."
1283,[xx:compliments],"[{'lang': 'main', 'text': 'Cottage Cheese Dry ..."
1285,[xx:compliments],"[{'lang': 'main', 'text': 'Lean Ground Pork'},..."
1288,[compliments],"[{'lang': 'main', 'text': 'Fruity Hoops'}, {'l..."


In [27]:
compliments_df['product_name'].dtype

dtype('O')

In [29]:
# Obtain the exact product name from 'main' language
def get_main_product_name(prod_name):
    if prod_name is None:
        return None
    for item in prod_name:
        if item['lang'] == 'main':
            return item['text']
    return None

In [31]:
# Obtain the exact product name from 'main' language where countries_tags is null
null_countries_products = compliments_df.loc[compliments_df['countries_tags'].isnull(), 'product_name'].apply(get_main_product_name)
null_countries_products

240                    Prepared horseradish
429                 Whole Wheat Naan Rounds
1273    Granulated Sweetener With Sucralose
1283                Cottage Cheese Dry Curd
1285                       Lean Ground Pork
1288                           Fruity Hoops
Name: product_name, dtype: object

In [33]:
compliments_df[compliments_df['product_quantity_unit'].isnull() & compliments_df['product_quantity'].isnull() & compliments_df['quantity'].notnull()]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
2,[xx:compliments],Compliments,0055742350500,[en:france],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...","[en:peanut, fr:sucre-a-glacer, en:hydrogenated...","[en:peanut, en:nut, fr:sucre-a-glacer, en:hydr...","[{""percent_max"":100.0,""percent_min"":20.0,""is_i...","[{'name': 'energy', 'value': None, '100g': 270...","[{'lang': 'main', 'text': 'Beurre d'arachide (...",None,None,,[],[]
11,[compliments],Compliments,0055742373561,[en:canada],"[{'key': 'ingredients_fr', 'imgid': 1.0, 'rev'...",None,None,None,None,"[{'lang': 'main', 'text': 'Brown Rice'}, {'lan...",None,None,,[],[]
17,[xx:compliments],Compliments,0055742500486,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'proteins', 'value': None, '100g': 1...","[{'lang': 'main', 'text': 'Farine sans gluten'...",None,None,,[],[]
26,[xx:compliments],Compliments,0055742508925,[en:france],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",[en:long-grain-rice],"[en:long-grain-rice, en:rice]","[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'energy', 'value': None, '100g': 148...","[{'lang': 'main', 'text': 'Riz à grain long'},...",None,None,,[],[]
55,[xx:compliments],Compliments,0068820121701,[en:france],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...","[fr:poulet-separe-mecaniquement, en:wheat-flou...","[fr:poulet-separe-mecaniquement, en:wheat-flou...","[{""percent_max"":100.0,""percent_min"":6.66666666...","[{'name': 'added-sugars', 'value': None, '100g...","[{'lang': 'main', 'text': 'Saucisses fumees au...",None,None,,[],[]
58,[xx:compliments],Compliments,0623682103588,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 0.0, ...","[{'lang': 'main', 'text': 'Jus D'orange (sans ...",None,None,,[],[]
59,[xx:compliments],Compliments,0623682105766,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 0.020...","[{'lang': 'main', 'text': 'Sirop D'erable Pur'...",None,None,,[],[]
63,[xx:compliments],Compliments,0055742528404,[en:france],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...","[en:fortified-wheat-flour, en:oat-flakes, en:p...","[en:fortified-wheat-flour, en:cereal, en:flour...","[{""percent_max"":100.0,""percent_min"":6.66666666...","[{'name': 'pantothenic-acid', 'value': None, '...","[{'lang': 'main', 'text': 'Oatmeal chocolate c...",None,None,,[],[]
97,[xx:compliments],Compliments,0055742357509,"[en:canada, en:france]","[{'key': '2', 'imgid': None, 'rev': None, 'siz...",[en:large-chicken-eggs],[en:large-chicken-eggs],"[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'fruits-vegetables-nuts', 'value': N...","[{'lang': 'en', 'text': 'Large white eggs'}, {...",None,None,12 pcs,[],[]
116,[xx:compliments],Compliments,0055742371314,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...","[fr:corn-starch, fr:carrot, fr:canola-oil, fr:...","[fr:corn-starch, fr:carrot, fr:canola-oil, fr:...","[{""percent_max"":100.0,""percent_min"":2.85714285...","[{'name': 'energy-kcal', 'value': None, '100g'...","[{'lang': 'main', 'text': 'Lasagnes'}, {'lang'...",None,None,,[],[]


In [35]:
compliments_df['quantity'].unique()

array(['300 g', '500 g', '', '900g', None, '796 mL', '600g', '350 g',
       '341 ml', '398 mL', '250g', '6 x 113 g', '350g', '170g', '500 mL',
       '1 L', '398 ml', '408 g', '336 g', '200 ml', '48 mL', '946 ml',
       '1 kg', '2.25kg', '200 g', '170 g', '4,54 kg', '210 g', '227g',
       '700 ml', '280g', '360g', '1,5 l', '225g', '300g', '300 ml',
       '284 mL', '1kg', '100g', '490 g', '540 gr', '280 g', '320gr',
       '400g', '12 pcs', '10g', '225 g', '200g', '2kg', '400 grammes',
       '375 ml', '284 ml', '450 g', '675 g', '473 mL', '340mL', '355 mL',
       '145 g', '375 mL', '340 g', '6 x 113 g,  678 g', '540 mL', '100 g',
       '10 Tortillas, 610 g', '2 L', '890 mL', '6x 75 g,  Net: 450 g',
       '454 g', '85 g', '250 g', '40 g', '8 x 142 g', '1.5 L', '946 mL',
       '1', '454g', '2.5 kg', '6 x 170 g', '2 kg', '750g', '390g',
       '900 mL', '16 rounds, 480 g.', '650 mL', '503 mL', '6 x 6 oz',
       '700 g', '473 ml', '12 x 90 g', '3 x 80 g', '342 g /  6 portions',
  

## Data Cleaning

### Brands_Tags

In [37]:
# Replace all the inconsistent values ['xx:Compliments', 'compliments'] with a standard value ['xx:compliments']
def normalize_brand_tags(tags):
    normalized_brand_tag = []
    for tag in tags:
        #split the brand_tag by ':' and convert it to lowercase
        brand = tag.split(":", 1)[-1].lower()
        normalized_brand_tag.append(f'xx:{brand}')
    return normalized_brand_tag

In [39]:
# Filter out brands_tags which are not ['xx:compliments']
inconsistent_tag = compliments_df['brands_tags'].apply(lambda tags: any(t != f'xx:{t.split(":", 1)[-1].lower()}' for t in tags))

# Apply normalization function to only those tags which are inconsistent
compliments_df.loc[inconsistent_tag, 'brands_tags'] = compliments_df.loc[inconsistent_tag, 'brands_tags'].apply(normalize_brand_tags)

In [41]:
# Verify whether the normalization has been applied
compliments_df['brands_tags'].explode().unique()

array(['xx:compliments'], dtype=object)

### Brands

In [44]:
compliments_df['brands'].unique()

array(['Compliments'], dtype=object)

In [46]:
# Check the distribution of each data type in the column 'brands'
compliments_df['brands'].apply(type).value_counts()

brands
<class 'str'>    1835
Name: count, dtype: int64

In [50]:
# Convert the datatype to string
compliments_df['brands'] = compliments_df['brands'].str.lower()

In [52]:
#Verify whether the normalization has been applied
compliments_df['brands'].unique()

array(['compliments'], dtype=object)

### Countries_Tags

In [54]:
#Find the unique values of countries_tags
compliments_df['countries_tags'].explode().unique()

array(['en:canada', 'en:france', 'en:saint-pierre-and-miquelon',
       'en:united-states', nan, None, 'en:germany', 'en:italy',
       'en:ireland', 'en:switzerland', 'en:austria', 'en:united-kingdom',
       'en:japan'], dtype=object)

In [56]:
# Normalize countries_tags to ensure all the products are sold in Canada
compliments_df['countries_tags'] = [['en:canada']]*len(compliments_df)

In [58]:
compliments_df['countries_tags'].explode().unique()

array(['en:canada'], dtype=object)

### Product_Name

In [60]:
len(compliments_df)

1835

In [62]:
compliments_df['product_name'][118]

array([], dtype=object)

In [64]:
for i in range(len(compliments_df)):
    if len(compliments_df['product_name'][i]) != 2:
        print(f"{i} : {len(compliments_df['product_name'][i])}")

0 : 3
6 : 3
13 : 3
22 : 3
23 : 3
27 : 3
31 : 3
32 : 3
33 : 3
34 : 3
36 : 3
37 : 3
39 : 3
40 : 3
43 : 3
45 : 1
47 : 3
50 : 3
53 : 3
62 : 3
65 : 3
70 : 3
71 : 3
76 : 3
77 : 1
79 : 3
80 : 3
85 : 3
88 : 3
91 : 3
92 : 3
102 : 3
104 : 3
111 : 3
118 : 0
119 : 3
122 : 3
129 : 3
130 : 0
134 : 3
135 : 3
138 : 3
139 : 3
144 : 0
145 : 3
148 : 3
150 : 0
152 : 3
154 : 3
156 : 3
157 : 0
160 : 3
163 : 3
173 : 3
175 : 3
176 : 3
179 : 3
183 : 3
184 : 3
185 : 3
188 : 3
189 : 3
190 : 3
193 : 3
194 : 3
195 : 3
196 : 1
197 : 3
198 : 3
205 : 3
206 : 3
208 : 3
211 : 3
213 : 3
216 : 3
217 : 3
218 : 3
221 : 3
223 : 3
224 : 3
228 : 3
230 : 3
232 : 3
233 : 3
239 : 3
242 : 3
243 : 3
245 : 3
251 : 3
252 : 0
257 : 3
262 : 3
265 : 3
266 : 0
270 : 3
274 : 3
278 : 3
282 : 3
286 : 3
287 : 3
292 : 3
294 : 3
297 : 3
298 : 3
299 : 3
302 : 3
303 : 3
304 : 3
307 : 3
308 : 3
311 : 3
315 : 0
316 : 3
319 : 3
321 : 3
325 : 0
331 : 3
332 : 3
333 : 3
337 : 3
338 : 3
339 : 3
343 : 3
345 : 3
352 : 3
354 : 0
362 : 3
373 : 3
374 : 3
3

In [66]:
compliments_df[compliments_df['product_name'].apply(len) == 0]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
118,[xx:compliments],compliments,0055742359657,[en:canada],[],None,None,None,None,[],None,None,None,[],[]
130,[xx:compliments],compliments,0068820132448,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,[],None,None,None,[],[]
144,[xx:compliments],compliments,0055742539776,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,[],None,None,None,[],[]
150,[xx:compliments],compliments,0055742346282,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,[],None,None,None,[],[]
157,[xx:compliments],compliments,0055742374629,[en:canada],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...","[en:water, en:tomato-concentrate, en:salt, en:...","[en:water, en:tomato-concentrate, en:vegetable...","[{""percent_max"":100.0,""percent_min"":47.0,""is_i...","[{'name': 'energy-kcal', 'value': None, '100g'...",[],None,None,,[],[]
252,[xx:compliments],compliments,0623682120622,[en:canada],"[{'key': 'front_en', 'imgid': 3.0, 'rev': 7.0,...",None,None,None,None,[],None,None,None,[],None
266,[xx:compliments],compliments,0055742531688,[en:canada],"[{'key': 'ingredients_en', 'imgid': 1.0, 'rev'...",None,None,None,None,[],None,None,None,[],None
315,[xx:compliments],compliments,0055742500547,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",[en:enriched-wheat-flour-water-canola-and-soyb...,[en:enriched-wheat-flour-water-canola-and-soyb...,"[{""percent_max"":100.0,""percent_min"":16.6666666...",[{'name': 'fruits-vegetables-legumes-estimate-...,[],None,None,None,[],[]
325,[xx:compliments],compliments,0055742531282,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...","[en:hot-banana-peppers-water-vinegar-salt, en:...","[en:hot-banana-peppers-water-vinegar-salt, en:...","[{""percent_max"":100.0,""percent_min"":50.0,""is_i...","[{'name': 'fruits-vegetables-nuts', 'value': N...",[],None,None,None,[],[]
354,[xx:compliments],compliments,0055742373257,[en:canada],"[{'key': 'packaging_fr', 'imgid': 4.0, 'rev': ...",None,None,None,None,[],None,240,3 x 80 g,[],None


### Observations:
- 0068820132448 = Cultivar Tomatoes Pack
- 0055742539776 = Sliceez Cheddar Style
- 0055742346282 = Bread and Butter Pickles
- 0055742374629 = Back of the Can with only ingredients visible
- 0623682120622 = Sensations Luxury biscuit collection (not a Compliments product)
- 0055742531688 = No image
- 0055742500547 = Back of the pack with ingredients visible
- 0055742531282 = Pickled hot peppers
- 0055742373257 = Butter Flavor Microwave Popcorn
- 0055742570588 = Lime Juice Bottle
- 0068820125570 = Naturally Smoked Bacon
- 0055742371338 = Fruit Cocktail Can
- 0055742505702 = Chopped Spinach packet
- 0055742386165 = Whole White Mushrooms
- 0068820125976 = Baby Spinach Packet
- 0055742524277 = Back of the bottle with ingredients visible
- 0055742546859 = Natural Almonds

### Images

In [68]:
compliments_df['images'].apply(type).value_counts()

images
<class 'numpy.ndarray'>    1835
Name: count, dtype: int64

In [70]:
# Find rows where there's no image ([])
compliments_df[compliments_df['images'].apply(lambda x: len(x) == 0)]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
18,[xx:compliments],compliments,0055742500752,[en:canada],[],"[en:rice, en:sugar, en:salt]","[en:rice, en:sugar, en:added-sugar, en:disacch...","[{""percent_max"":100.0,""percent_min"":92.949,""is...",[{'name': 'fruits-vegetables-nuts-estimate-fro...,"[{'lang': 'main', 'text': 'Gluten-Free Crisp R...",None,None,None,[],[]
111,[xx:compliments],compliments,0055742510423,[en:canada],[],None,None,None,"[{'name': 'potassium', 'value': None, '100g': ...","[{'lang': 'main', 'text': 'Mais congelé'}, {'l...",g,500.0,500 g,[],[]
118,[xx:compliments],compliments,0055742359657,[en:canada],[],None,None,None,None,[],None,None,None,[],[]
121,[xx:compliments],compliments,0055742359688,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Amandes émondées'},...",g,100,100g,[],[]
122,[xx:compliments],compliments,0055742360196,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Maïs en crème'}, {'...",ml,284,284 ml,[],[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1828,[xx:compliments],compliments,0220102515961,[en:canada],[],None,None,None,"[{'name': 'iron', 'value': None, '100g': 0.0, ...","[{'lang': 'main', 'text': 'Chicken Breast'}, {...",None,None,None,[],None
1829,[xx:compliments],compliments,0055742373370,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Fajita Seasoning Mi...",None,None,None,[],None
1831,[xx:compliments],compliments,0270087010555,[en:canada],[],None,None,None,"[{'name': 'fat', 'value': None, '100g': 2.0, '...","[{'lang': 'main', 'text': 'Oven Roasted Pepper...",None,None,None,[],None
1832,[xx:compliments],compliments,0055742584110,[en:canada],[],None,None,None,"[{'name': 'trans-fat', 'value': None, '100g': ...","[{'lang': 'main', 'text': 'Grilling Sausage Sm...",None,None,None,[],None


### Observation:
- 355 products out of 1835 don't have an image. Most of them don't have corresponding ingredients as well. Although a few of them have ingredients, we can't ensure that the information is accurate because of the missing image.
- Since we can't infer any missing information without the product image, they are safe to be removed to maintain accuracy.

### Ingredients

In [72]:
compliments_df['ingredients_original_tags'].apply(type).value_counts()

ingredients_original_tags
<class 'NoneType'>         1363
<class 'numpy.ndarray'>     472
Name: count, dtype: int64

In [74]:
compliments_df[compliments_df['ingredients_original_tags'].apply(lambda x: x is None)]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
4,[xx:compliments],compliments,0055742358988,[en:canada],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Melasse'}, {'lang':...",None,None,None,[],[]
6,[xx:compliments],compliments,0055742363425,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 0.620...","[{'lang': 'main', 'text': 'Crushed tomatoes'},...",ml,796.0,796 mL,[],[]
11,[xx:compliments],compliments,0055742373561,[en:canada],"[{'key': 'ingredients_fr', 'imgid': 1.0, 'rev'...",None,None,None,None,"[{'lang': 'main', 'text': 'Brown Rice'}, {'lan...",None,None,,[],[]
12,[xx:compliments],compliments,0055742374018,[en:canada],"[{'key': '5', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'carbohydrates', 'value': None, '100...","[{'lang': 'main', 'text': 'Low salt whole kern...",ml,341,341 ml,[],[]
14,[xx:compliments],compliments,0055742380705,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'fat', 'value': None, '100g': 10.0, ...","[{'lang': 'main', 'text': 'Stone ground wheat ...",None,None,None,[],[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1830,[xx:compliments],compliments,0055742511970,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 1.833...","[{'lang': 'main', 'text': 'Light Mayonnaise'},...",None,None,None,[],None
1831,[xx:compliments],compliments,0270087010555,[en:canada],[],None,None,None,"[{'name': 'fat', 'value': None, '100g': 2.0, '...","[{'lang': 'main', 'text': 'Oven Roasted Pepper...",None,None,None,[],None
1832,[xx:compliments],compliments,0055742584110,[en:canada],[],None,None,None,"[{'name': 'trans-fat', 'value': None, '100g': ...","[{'lang': 'main', 'text': 'Grilling Sausage Sm...",None,None,None,[],None
1833,[xx:compliments],compliments,0055742581621,[en:canada],[],None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Mini Butter Tarts'}...",None,None,None,[],None


In [76]:
compliments_df[compliments_df['images'].apply(lambda x: len(x) != 0) & compliments_df['ingredients_original_tags'].apply(lambda x: x is None)]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
4,[xx:compliments],compliments,0055742358988,[en:canada],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Melasse'}, {'lang':...",None,None,None,[],[]
6,[xx:compliments],compliments,0055742363425,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 0.620...","[{'lang': 'main', 'text': 'Crushed tomatoes'},...",ml,796.0,796 mL,[],[]
11,[xx:compliments],compliments,0055742373561,[en:canada],"[{'key': 'ingredients_fr', 'imgid': 1.0, 'rev'...",None,None,None,None,"[{'lang': 'main', 'text': 'Brown Rice'}, {'lan...",None,None,,[],[]
12,[xx:compliments],compliments,0055742374018,[en:canada],"[{'key': '5', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'carbohydrates', 'value': None, '100...","[{'lang': 'main', 'text': 'Low salt whole kern...",ml,341,341 ml,[],[]
14,[xx:compliments],compliments,0055742380705,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'fat', 'value': None, '100g': 10.0, ...","[{'lang': 'main', 'text': 'Stone ground wheat ...",None,None,None,[],[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1811,[xx:compliments],compliments,0055742559064,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Jalapeno havarti ch...",None,None,None,[],None
1813,[xx:compliments],compliments,0055742578102,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Vegan protein+ gree...",None,None,None,[],None
1826,[xx:compliments],compliments,0055742582048,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'proteins', 'value': None, '100g': 0...","[{'lang': 'main', 'text': 'Frozt Bitez'}, {'la...",None,None,None,[],None
1830,[xx:compliments],compliments,0055742511970,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 1.833...","[{'lang': 'main', 'text': 'Light Mayonnaise'},...",None,None,None,[],None


### Nutriments

In [78]:
compliments_df[compliments_df['nutriments'].apply(lambda x: x is None)]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
11,[xx:compliments],compliments,0055742373561,[en:canada],"[{'key': 'ingredients_fr', 'imgid': 1.0, 'rev'...",None,None,None,None,"[{'lang': 'main', 'text': 'Brown Rice'}, {'lan...",None,None,,[],[]
15,[xx:compliments],compliments,0055742380712,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,"[{'lang': 'main', 'text': 'Craquelins de blé m...",None,250,250g,[],[]
44,[xx:compliments],compliments,0055742539004,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,"[{'lang': 'main', 'text': 'Quick oats'}, {'lan...",None,2250,2.25kg,[],[]
57,[xx:compliments],compliments,01070194,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,"[{'lang': 'main', 'text': 'Organic Smooth Pean...",None,500,500 g,[],[]
60,[xx:compliments],compliments,0623682114737,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,None,"[{'lang': 'main', 'text': 'Macadamia madness'}...",None,None,None,[],[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1805,[xx:compliments],compliments,0055742581140,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Quinoa'}, {'lang': ...",None,None,None,[],None
1810,[xx:compliments],compliments,0055742351767,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Bread & Butter Pick...",None,None,None,[],None
1812,[xx:compliments],compliments,0057316211385,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'peas and carrots'},...",None,None,None,[],None
1821,[xx:compliments],compliments,0779566119760,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Sole Mio'}, {'lang'...",None,None,None,[],None


### Observation: 
- 0055742359657 = Doesn't have any information except barcode

In [81]:
compliments_df[compliments_df['images'].apply(lambda x: len(x) == 0) & compliments_df['nutriments'].apply(lambda x: x is None)]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
118,[xx:compliments],compliments,0055742359657,[en:canada],[],None,None,None,None,[],None,None,None,[],[]
121,[xx:compliments],compliments,0055742359688,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Amandes émondées'},...",g,100,100g,[],[]
122,[xx:compliments],compliments,0055742360196,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Maïs en crème'}, {'...",ml,284,284 ml,[],[]
453,[xx:compliments],compliments,0055742552072,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Café'}, {'lang': 'e...",None,None,None,[],None
789,[xx:compliments],compliments,0055742509373,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Sacs à collations'}...",None,None,75,[],None
1391,[xx:compliments],compliments,0055742375336,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Apple Cider Vinegar...",None,0,1pcs,[],None
1752,[xx:compliments],compliments,0055742570687,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': '100% Pure Coconut W...",None,None,None,[],None
1771,[xx:compliments],compliments,0068820134510,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Cranberry Lemon Muf...",None,None,None,[],None
1787,[xx:compliments],compliments,0055742570427,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': '100% Pure Coconut W...",None,None,None,[],None
1798,[xx:compliments],compliments,0055742581928,[en:canada],[],None,None,None,None,"[{'lang': 'main', 'text': 'Deluxe Mixed Nut'},...",None,None,None,[],None


In [83]:
len(compliments_df[compliments_df['images'].apply(lambda x: len(x) == 0) & compliments_df['nutriments'].apply(lambda x: x is None)])

15

### Observation:
- There are 15 products which contains a lot of missing information. The only information we have for these products is code and product_name.
- Since we can't infer anything from the existing fields and fill in the missing fields, these rows are safe to be removed to maintain accurate data.
- There are 62 products where there is an image. We can explore the possibility of filling the nutriments value using the product image.

### Quantity

In [85]:
compliments_df['quantity'].apply(type).value_counts()

quantity
<class 'NoneType'>    1324
<class 'str'>          511
Name: count, dtype: int64

In [87]:
#Find the rows where quantity is None
compliments_df[compliments_df['quantity'].isna()]

,brands_tags,brands,code,countries_tags,images,ingredients_original_tags,ingredients_tags,ingredients,nutriments,product_name,product_quantity_unit,product_quantity,quantity,traces_tags,vitamins_tags
4,[xx:compliments],compliments,0055742358988,[en:canada],"[{'key': '3', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Melasse'}, {'lang':...",None,None,None,[],[]
5,[xx:compliments],compliments,0055742359886,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...","[fr:not-a-significant-source-of-saturated-fat,...","[fr:not-a-significant-source-of-saturated-fat,...","[{""percent_max"":100.0,""percent_min"":6.25,""is_i...","[{'name': 'added-sugars', 'value': None, '100g...","[{'lang': 'main', 'text': 'Persil Déshydraté E...",None,None,None,[],[en:vitamin-a]
8,[xx:compliments],compliments,0055742367164,[en:canada],"[{'key': '4', 'imgid': None, 'rev': None, 'siz...",[fr:peanuts],[fr:peanuts],"[{""percent_max"":100.0,""percent_min"":100.0,""is_...","[{'name': 'fat', 'value': None, '100g': 46.666...","[{'lang': 'main', 'text': 'Natural crunchy pea...",None,None,None,[],[]
14,[xx:compliments],compliments,0055742380705,[en:canada],"[{'key': '2', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'fat', 'value': None, '100g': 10.0, ...","[{'lang': 'main', 'text': 'Stone ground wheat ...",None,None,None,[],[]
18,[xx:compliments],compliments,0055742500752,[en:canada],[],"[en:rice, en:sugar, en:salt]","[en:rice, en:sugar, en:added-sugar, en:disacch...","[{""percent_max"":100.0,""percent_min"":92.949,""is...",[{'name': 'fruits-vegetables-nuts-estimate-fro...,"[{'lang': 'main', 'text': 'Gluten-Free Crisp R...",None,None,None,[],[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1830,[xx:compliments],compliments,0055742511970,[en:canada],"[{'key': '1', 'imgid': None, 'rev': None, 'siz...",None,None,None,"[{'name': 'salt', 'value': None, '100g': 1.833...","[{'lang': 'main', 'text': 'Light Mayonnaise'},...",None,None,None,[],None
1831,[xx:compliments],compliments,0270087010555,[en:canada],[],None,None,None,"[{'name': 'fat', 'value': None, '100g': 2.0, '...","[{'lang': 'main', 'text': 'Oven Roasted Pepper...",None,None,None,[],None
1832,[xx:compliments],compliments,0055742584110,[en:canada],[],None,None,None,"[{'name': 'trans-fat', 'value': None, '100g': ...","[{'lang': 'main', 'text': 'Grilling Sausage Sm...",None,None,None,[],None
1833,[xx:compliments],compliments,0055742581621,[en:canada],[],None,None,None,"[{'name': 'saturated-fat', 'value': None, '100...","[{'lang': 'main', 'text': 'Mini Butter Tarts'}...",None,None,None,[],None
